src/Workspaces/Core/Portable/Formatting/Engine/TokenStream.cs

```diff
@@ -501,9 +530,35 @@ private bool IsFirstTokenOnLine(TokenData tokenData1, TokenData tokenData2)
             return this.GetTriviaData(tokenData1, tokenData2).SecondTokenIsFirstTokenOnLine;
         }
 
+        /// <summary>
+        /// An approximation for the length of a token in this stream
+        /// </summary>
+        float AverageTokenLength { get; }
+        /// <summary>
+        /// Gets approximated index bounds for a token with specified position, which will definitely contain
+        /// the token, if it is present in the stream.
+        /// </summary>
+        private void GetTokenIndexBounds(int position, out int lowerBound, out int upperBound)
+        {
+            float avgLength = this.AverageTokenLength;
+            Debug.Assert(avgLength > 0);
+            lowerBound = upperBound = Bounded((int)((position - _tokens[0].Position) / avgLength), min: 0, max: _tokens.Length - 1);
+
+            while (_tokens[lowerBound].Position >= position && lowerBound > 0)
+            {
+                lowerBound = Math.Max(0, lowerBound - Math.Max(1, (int)((_tokens[lowerBound].Position - position) / avgLength)));
+            }
+
+            while (_tokens[upperBound].Position <= position && upperBound < _tokens.Length - 1)
+            {
+                upperBound = Math.Min(_tokens.Length - 1, upperBound + Math.Max(1, (int)((position - _tokens[upperBound].Position) / avgLength)));
+            }
+        }
```

Comment:
please docuemnt what this method is doing and why it's correct. it's not trivial to understand or validate.

Reply:
(skipping for now).

Reply:
Added a couple of comments.

Reply:
comments very helpful, thanks.